### DISCUSSION TITLE
x-text within x-for not triggered after x-data changes..

### DISCUSSION THREAD

**Discussion by pjbakker** [alpinejs/alpine#1406] [7. Bugs (archived)] [answered]
## What is intended

In the below code, the intention is to:

* Add an uploaded filename to a visible list, setting the status to 'Uploading' and displaying it. **(Works)**
* trigger an upload after a file is dropped onto a file input **(Works)**
* Update the status of the upload when it does, and show the result to the user **(Update not triggered)**

## What should happen

When the div with `@upload-result.window` receives the event and updates the relevant `file` in the `files array` (from the x-data), the monitoring `<span x-text='files[index].result>` should update the visual text on the span.

## What happens

Although the value in `files[index].result` is updated, which can be seen when logging to console, the actual visual text in the `span` is not.

## Workaround

By adding `files.push([]); files.pop;` (see below) in the `@upload-result.window` event handler, it does trigger an update of the visual text on the span.

## Code

~~~
      <div class="mt-6 max-w-lg space-y-6">
        <div x-data="upload()" @upload-result.window="files[$event.detail.index].result = $event.detail.result; files.push([]); files.pop();" class="flex flex-col flex-grow mb-3">
          <label for="cover_photo" class="block text-sm font-medium text-gray-700">
            Attachments
          </label>
          <div x-ref="border" id="FileUpload" class="block w-full py-2 px-3 relative bg-white appearance-none border-2 border-gray-300 border-dashed rounded-md hover:shadow-outline-gray ">
            <input type="file"
               class="absolute inset-0 z-50 m-0 p-0 w-full h-full outline-none opacity-0"
               x-on:change="start($event.target.files, $dispatch)"
            >
            <div class="flex flex-col space-y-2 items-center justify-center">
              <svg class="mx-auto h-12 w-12 text-gray-400" stroke="currentColor" fill="none" viewBox="0 0 48 48" aria-hidden="true">
                <path d="M28 8H12a4 4 0 00-4 4v20m32-12v8m0 0v8a4 4 0 01-4 4H12a4 4 0 01-4-4v-4m32-4l-3.172-3.172a4 4 0 00-5.656 0L28 28M8 32l9.172-9.172a4 4 0 015.656 0L28 28m0 0l4 4m4-24h8m-4-4v8m-12 4h.02" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" />
              </svg>
              <div class="flex text-sm text-gray-600">
                <a href="javascript:void(0)" class="cursor-pointer bg-white rounded-md font-medium text-indigo-600 hover:text-indigo-500 focus-within:outline-none focus-within:ring-2 focus-within:ring-offset-2 focus-within:ring-indigo-500">Upload a file</a>
              </div>
            </div>
          </div>
          <template x-if="files !== null">
            <div class="flex flex-col space-y-1">
              <template x-for="(_,index) in Array.from({ length: files.length })">
                <div class="flex flex-row items-center space-x-2 text-sm font-medium text-gray-700">
                  <span x-text="files[index].name"></span>
                  <span x-text="files[index].result"></span>
                </div>
              </template>
            </div>
          </template>
        </div>
      </div>

<script>
function upload() {
    return {
        files: [],
        start(uploadedFiles, dispatch) {
            for (i = 0; i < uploadedFiles.length; i++) {
                let file = uploadedFiles[i];
                file.result = 'Starting';

                let index = this.files.length;
                this.files.push(file);
                file.result = 'Uploading';

                let formData = new FormData();
                formData.append('file', file);

                fetch('/upload',
                  { method: 'POST', body: formData })
                  .then( async (response) => {
                          let data = await response.json();

                          if (response.status === 200) {
                            dispatch('upload-result', { index: index, result: 'success'} );
                          } else {
                            dispatch('upload-result', { index: index, result: 'failed' } );
                          }
                  })
                  .catch(function (error) {
                          dispatch('upload-result', { index: index, result: 'exception' } );
                  });
            }
        },
    }
}
</script>
~~~

**Comment by KevinBatdorf** [comment]
I might not be 100% correct on this, but I believe the mechanism which monitors the component state would need to be aware of `files` property in order to monitor it for changes, so you would need the following to be known beforehand, which obviously isn't ideal at all.

```js
{
  files: {
    0: {
      result: null,
    },
}
```
You're better off updating the state another way. Either your workaround, or perhaps:
```html
<div @upload-result.window="files[$event.detail.index].result = $event.detail.result;$el.__x.updateElements($el)">
```
(That's an internal method though so could the implementation could of course change without notice)

**Reply by pjbakker** [reply]
Doesn't work anymore in Alpine v3 as `__x` does not exist anymore.

**Comment by pjbakker** [comment]
I thought something like that as well, but `x-text` correctly monitors and updates the initial change of `file.result` from `Starting` to `Uploading`. And, when I do anything else that triggers an update, it does change.

> You're better off updating the state another way. Either your workaround, or perhaps:
> 
> ```
> <div @upload-result.window="files[$event.detail.index].result = $event.detail.result;$el.__x.updateElements($el)">
> ```
> 
> (That's an internal method though so could the implementation could of course change without notice)

Internal but much cleaner, thanks!

**Comment by HugoDF** [comment]
> I thought something like that as well, but `x-text` correctly monitors and updates the initial change of `file.result` from `Starting` to `Uploading`. And, when I do anything else that triggers an update, it does change.
> 
> > You're better off updating the state another way. Either your workaround, or perhaps:
> > ```
> > <div @upload-result.window="files[$event.detail.index].result = $event.detail.result;$el.__x.updateElements($el)">
> > ```
> > 
> > 
> > (That's an internal method though so could the implementation could of course change without notice)
> 
> Internal but much cleaner, thanks!

alternate ways to trigger the update by cloning the `files` and re-assigning it as state or updating any other variable eg. `update = Date.now()`

**Reply by pjbakker** [reply]
This doesn't seem to work in Alpine v3 as the DOM tree tree is not walked anymore..

**Comment by pjbakker** [comment]
I can't be the only one that wants to update an upload status (or progress bar) based on events from a XmlHttpRequest (or fetch, or anything else)?

How can I trigger that the element is updated. I read about reactive() and effect() but cannot figure out how it would help in this case

**Reply by SimoTod** [reply]
V3 deep tracks the component variables so you should be able to do it without any workarounds.

**Reply by pjbakker** [reply]
I hoped that.. But this fiddle shows that it's not the case in this situation..

I'd love to see how to fix this..

https://jsfiddle.net/k0s37jxz/

**Reply by ekwoka** [reply]
Not sure exactly what is causing the issue but the way this is written is not typical for alpine.

You data methods should just be `thing() {}` and then for the callbacks passed to things, use anonymous arrow functions to maintain the this context.

Additionally, use `:key` on x-for

**Reply by SimoTod** [reply, marked as answer]
> I hoped that.. But this fiddle shows that it's not the case in this situation..
> 
> I'd love to see how to fix this..
> 
> https://jsfiddle.net/k0s37jxz/

@pjbakker Only literal objects and arrays are trackable. You are pushing a File class so the inner properties are not reactive.
One simple change would be `this.files.push(Object.assign({}, file));` instead of `this.files.push(file);` unless you need to get the real File later on from `this.files` (in that case, you need a separate array to store the tracking properties).

**Reply by ekwoka** [reply]
Ah yes that would do it


I normally do, for files, it make a wrapper object. So an object with `file` as a prop and my state thing (uploaded etc) on that object.

**Reply by pjbakker** [reply]
Thanks a million.. The object assignment fixes it.

I'm not a native JS dev. And sometimes these kind of different ways of thinking about objects vs associative arrays / hashes, blindsides me...
